File: typescript/day4/part2.ts
const inputFile = 'input.txt'

const puzzleGrid = (await Deno.readTextFile(inputFile))
	.split('\n')
	.filter(line => line.length > 0)
	.map(line => line.split(''))

const matchAt = (
	at: { x: number; y: number },
): 0 | 1 => {
	if (puzzleGrid[at.y][at.x] !== 'A') {
		return 0
	}

	const cornerValues = [
		puzzleGrid[at.y - 1][at.x + 1],
		puzzleGrid[at.y + 1][at.x + 1],
		puzzleGrid[at.y + 1][at.x - 1],
		puzzleGrid[at.y - 1][at.x - 1],
	]

	const numM = cornerValues.filter(char => char === 'M').length
	const numS = cornerValues.filter(char => char === 'S').length
	if (numM !== 2 || numS !== 2 || cornerValues[0] === cornerValues[2]) {
		return 0
	}

	return 1
}

let numAnswers = 0
for (let rowIndex = 1; rowIndex < puzzleGrid.length -1; ++rowIndex) {
	for (let columnIndex = 1; columnIndex < puzzleGrid[rowIndex].length -1; ++columnIndex) {
		numAnswers += matchAt({ x: columnIndex, y: rowIndex })
	}
}

console.log(`numAnswers = ${numAnswers}`)

File: rust/day4/src/part2.rs
use std::fs;


fn main() {
    let input_data = fs::read_to_string("input.txt").expect("Could not read input file.");
	let puzzle_data: Vec<Vec<char>> = input_data
		.lines()
		.filter(|line| !line.is_empty())
		.map(|line| line.chars().collect::<Vec<_>>())
		.collect::<Vec<_>>();

	let mut num_matches = 0;
	let num_rows = puzzle_data.len();
	// Assuming all rows are the same length
	let num_columns = puzzle_data[0].len();

	let matches_at = |row_index: usize, column_index: usize| -> bool {
		if puzzle_data[row_index][column_index] == 'A' {
			let corner_values = [
				puzzle_data[row_index - 1][column_index - 1],
				puzzle_data[row_index - 1][column_index + 1],
				puzzle_data[row_index + 1][column_index + 1],
				puzzle_data[row_index + 1][column_index - 1],
			];
			let num_m = corner_values.iter().filter(|&&c| c == 'M').count();
			let num_s = corner_values.iter().filter(|&&c| c == 'S').count();
			num_m == 2 && num_s == 2 && corner_values[0] != corner_values[2]
		} else {
			false
		}
	};

	for row_index in 1..num_rows - 1 {
		for column_index in 1..num_columns - 1 {
			if matches_at(row_index, column_index) {
				num_matches += 1;
			}
		}
	}

	println!("{num_matches}");
}
